pkg/storage/scrub.go

```diff
@@ -92,153 +91,234 @@ func CheckRepo(ctx context.Context, imageName string, imgStore storageTypes.Imag
+		imageResult := getManifestResult(ctx, manifest, imgStore, imageName, tag, scrubbedManifests)
+		results = append(results, imageResult)
+	}
 
-			var idx ispec.Index
-			if err := json.Unmarshal(buf, &idx); err != nil {
-				tagName := manifest.Annotations[ispec.AnnotationRefName]
-				imgRes := getResult(imageName, tagName, errors.ErrBadBlobDigest)
-				results = append(results, imgRes)
+	return results, nil
+}
 
-				continue
+func getManifestResult(
+	ctx context.Context, manifest ispec.Descriptor, imgStore storageTypes.ImageStore, imageName, tag string,
+	scrubbedManifests map[godigest.Digest]ScrubImageResult,
+) ScrubImageResult {
+	res, ok := scrubbedManifests[manifest.Digest]
+	if ok {
+		return ScrubImageResult{
+			ImageName:    imageName,
+			Tag:          tag,
+			Status:       res.Status,
+			AffectedBlob: res.AffectedBlob,
+			Error:        res.Error,
+		}
+	}
+
+	switch manifest.MediaType {
+	case ispec.MediaTypeImageIndex:
+		buf, err := imgStore.GetBlobContent(imageName, manifest.Digest)
+		if err != nil {
+			imgRes := getResult(imageName, tag, manifest.Digest, errors.ErrBadBlobDigest)
+			scrubbedManifests[manifest.Digest] = imgRes
+
+			return imgRes
+		}
+
+		var idx ispec.Index
+		if err := json.Unmarshal(buf, &idx); err != nil {
+			imgRes := getResult(imageName, tag, manifest.Digest, errors.ErrBadBlobDigest)
+			scrubbedManifests[manifest.Digest] = imgRes
+
+			return imgRes
+		}
+
+		for _, m := range idx.Manifests {
+			imgRes := getManifestResult(ctx, m, imgStore, imageName, tag, scrubbedManifests)
+			scrubbedManifests[manifest.Digest] = imgRes
+
+			if imgRes.Error != "" {
+				return imgRes
 			}
+		}
+
+		okResult := getResult(imageName, tag, "", nil)
+
+		scrubbedManifests[manifest.Digest] = okResult
 
-			listOfManifests = append(listOfManifests, idx.Manifests...)
-		} else if manifest.MediaType == ispec.MediaTypeImageManifest {
-			listOfManifests = append(listOfManifests, manifest)
+		// check subject if exists
+		if idx.Subject != nil {
+			imgRes := getManifestResult(ctx, *idx.Subject, imgStore, imageName, tag, scrubbedManifests)
+			scrubbedManifests[manifest.Digest] = imgRes
+
+			return imgRes
 		}
-	}
 
-	for _, m := range listOfManifests {
-		tag := m.Annotations[ispec.AnnotationRefName]
-		imageResult := CheckIntegrity(ctx, imageName, tag, oci, m, dir)
-		results = append(results, imageResult)
-	}
+		return okResult
+	case ispec.MediaTypeImageManifest:
+		imgRes := CheckIntegrity(ctx, imageName, tag, manifest, imgStore)
+		scrubbedManifests[manifest.Digest] = imgRes
 
-	return results, nil
+		// if integrity ok then check subject if exists
+		if imgRes.Error == "" {
+			manifestContent, _ := imgStore.GetBlobContent(imageName, manifest.Digest)
+
+			var m ispec.Manifest
+
+			_ = json.Unmarshal(manifestContent, &m)
+
+			if m.Subject != nil {
+				imgRes = getManifestResult(ctx, *m.Subject, imgStore, imageName, tag, scrubbedManifests)
+				scrubbedManifests[manifest.Digest] = imgRes
+
+				return imgRes
```

Comment:
Is this correct? We should be returning result for current image, not for the referred to image?

Reply:
We are returning the result for the current image based on the imageName and tag parameters sent to getManifestResult at 199.